### PR TITLE
Make ExpressionEvaluator support Double

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This would be needed for arange support, because for arange, then output tensor size is
`ceil((end - start)/step)`, where `end`, `start`, `step` could be floating points.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/dynamic_type.h [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+#include <c10/util/Exception.h>
+#include <cmath>
+#include <iostream>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class TORCH_CUDA_CU_API IntOrDouble {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/dynamic_type.h:14]: nit: can this be simplified with `c10::variant`?

File: torch/csrc/jit/codegen/cuda/dynamic_type.h [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  bool is_int() const {
+    return is_int_;
+  }
+
+  template <typename T>
+  T as() const;
+  template <typename T>
+  T cast() const;
+
+#define DEFINE_OP(op)                                       \
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/dynamic_type.h]: nit: use a more specific name for code readability

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 3]
```diff
@@ -193,16 +193,17 @@ bool Val::isConstScalar() const {
...
 bool Val::isConstInt() const {
-  if (!isAnInt()) {
-    return false;
-  }
-  return ConstCheck::isConstInt(this);
+  return isConstExpr() && isAnInt();
+}
+
+bool Val::isConstExpr() const {
+  return ConstCheck::isConst(this);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Do we need to separate it to the two functions?

The name of `Val::isConstExpr()` doesn't sound quite right as `Val` is not an expression.

**Review by naoyam — APPROVED** [review]
LGTM. Left some comments

File: torch/csrc/jit/codegen/cuda/dynamic_type.h (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/dynamic_type.h:14]: WOW, I was looking at `std::variant` and unfortunately, it was C++17 only. Didn't know that there is a `c10::variant`. Thanks for pointing this out.

File: torch/csrc/jit/codegen/cuda/dynamic_type.h (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/dynamic_type.h]: fixed

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Agree, removed

**Pull request merged**